packages/runtime/src/widgets/PopupMenu.tsx

```diff
@@ -38,99 +49,145 @@ interface PopupMenuStyles {
-        const value = evaluate<string | number>(
-          defaultScreenContext,
-          itemTemplate.value,
-          {
-            [itemTemplate.name]: get(item, itemTemplate.name) as unknown,
-          },
-        );
-
-        return (
-          <React.Fragment key={value}>
-            <Menu.Item key={value} onClick={() => action?.callback(item)}>
-              <CustomScopeProvider value={item as CustomScope}>
-                {EnsembleRuntime.render([itemTemplate.template])}
-              </CustomScopeProvider>
-            </Menu.Item>
-            {values?.showDivider
-              ? index < namedData.length - 1 && <Menu.Divider />
-              : null}
-          </React.Fragment>
-        );
+    if (isObject(itemTemplate) && !isEmpty(evaluatedNamedData.namedData)) {
+      const tempItems = evaluatedNamedData.namedData.map((item, index) => {
+        const itm: ItemType = {
+          key: `popupmenu_itemTemplate_${index}`,
+          label: (
+            <CustomScopeProvider value={item as CustomScope}>
+              {EnsembleRuntime.render([itemTemplate.template])}
+            </CustomScopeProvider>
+          ),
+        };
+        return itm;
       });
 
       popupItems.push(...tempItems);
     }
 
-    return (
-      <Menu className={values?.styles?.names} style={{ ...values?.styles }}>
-        {popupItems}
-      </Menu>
-    );
-  }, [values?.items, action, namedData, itemTemplate]);
+    if (values?.showDivider) {
+      for (let i = 1; i < popupItems.length; i += 2) {
+        popupItems.splice(i, 0, { type: "divider" });
+      }
+    }
+
+    return popupItems;
+  }, [
+    values?.items,
+    values?.showDivider,
+    itemTemplate,
+    evaluatedNamedData.namedData,
+    getMenuItem,
+  ]);
 
-  useEffect((): void => {
-    if (values && !values.widget) {
-      return;
+  const getWidget = useCallback(() => {
+    if (!values?.widget) {
+      throw Error("PopupMenu requires a widget to render the anchor.");
     }
-    // clone value so we're not updating the yaml doc
-    const widget = cloneDeep(values?.widget);
-    const actualWidget = unwrapWidget(widget!);
-    setWidgetProps(actualWidget);
-    // Only run once
-  }, []);
+    const widget = cloneDeep(values.widget);
+    const actualWidget = unwrapWidget(widget);
+    return EnsembleRuntime.render([actualWidget]);
+  }, [values?.widget]);
+
+  const handleMenuItemClick = useCallback<NonNullable<MenuProps["onClick"]>>(
+    ({ key }) => {
+      let item;
+      if (key.includes("itemTemplate")) {
+        const itemIndex = toNumber(key.at(-1));
+        item = evaluatedNamedData.namedData[itemIndex];
+      } else {
+        const itemIndices = key.split("_").filter((_, i) => i > 1);
+        if (itemIndices.length > 1) {
+          for (let i = 1; i < itemIndices.length; i += 2) {
+            itemIndices.splice(i, 0, "items");
+          }
+        }
+        item = get(values?.items, itemIndices) as unknown;
+      }
```

Comment:
Does this work if you have double digit items? e.g. 12 items.

Reply:
nice catch, ty.